frame/vesting/src/lib.rs

```diff
@@ -302,121 +407,322 @@ pub mod pallet {
+						"faulty logic led to attempting to add too many vesting schedules",
+					);
+					return Err(Error::<T>::AtMaxVestingSchedules.into());
+				}
+				Self::deposit_event(Event::<T>::VestingMergeSuccess(
+					s.locked(),
+					s.per_block(),
+					s.starting_block(),
+				));
+				Vesting::<T>::insert(&who, vesting);
+			} else if let Some(vesting) = maybe_vesting {
+				Vesting::<T>::insert(&who, vesting);
+			} else {
+				Vesting::<T>::remove(&who);
+			}
 
 			Ok(())
 		}
 	}
 }
 
 impl<T: Config> Pallet<T> {
+	// Create a new `VestingInfo`, based off of two other `VestingInfo`s.
+	// NOTE: We assume both schedules have been vested up through the current block.
+	fn merge_vesting_info(
+		now: T::BlockNumber,
+		schedule1: VestingInfo<BalanceOf<T>, T::BlockNumber>,
+		schedule2: VestingInfo<BalanceOf<T>, T::BlockNumber>,
+	) -> Option<VestingInfo<BalanceOf<T>, T::BlockNumber>> {
+		let schedule1_ending_block = schedule1.ending_block::<T::BlockNumberToBalance>();
+		let schedule2_ending_block = schedule2.ending_block::<T::BlockNumberToBalance>();
+		let now_as_balance = T::BlockNumberToBalance::convert(now);
+
+		// Check if one or both schedules have ended.
+		match (schedule1_ending_block <= now_as_balance, schedule2_ending_block <= now_as_balance) {
+			// If both schedules have ended, we don't merge and exit early.
+			(true, true) => return None,
+			// If one schedule has ended, we treat the one that has not ended as the new
+			// merged schedule.
+			(true, false) => return Some(schedule2),
+			(false, true) => return Some(schedule1),
+			// If neither schedule has ended don't exit early.
+			_ => {},
+		}
+
+		let locked = schedule1
+			.locked_at::<T::BlockNumberToBalance>(now)
+			.saturating_add(schedule2.locked_at::<T::BlockNumberToBalance>(now));
+		// This shouldn't happen because we know at least one ending block is greater than now.
+		if locked.is_zero() {
+			return None;
+		}
+
+		let ending_block = schedule1_ending_block.max(schedule2_ending_block);
+		let starting_block = now.max(schedule1.starting_block()).max(schedule2.starting_block());
+		let duration =
+			ending_block.saturating_sub(T::BlockNumberToBalance::convert(starting_block));
+		let per_block = if duration.is_zero() {
+			// The logic of `ending_block` guarantees that each schedule ends at least a block
+			// after it starts and since we take the max starting and ending_block we should never
+			// get here
+			locked
+		} else if duration > locked {
+			// This would mean we have a per_block of less than 1, which should not be not possible
+			// because when we create the new schedule is at most the same duration as the longest,
+			// but never greater.
+			1u32.into()
+		} else {
+			locked.checked_div(&duration)?
+		};
+
+		// At this point inputs have been validated, so this should always be `Some`.
+		let schedule = VestingInfo::new::<T>(locked, per_block, starting_block);
+		debug_assert!(schedule.validate::<T>().is_ok());
```

Comment:
I wanted to raise before as well that `validate` also alters the schedule right? this might create a discrepancy between test and release.

Reply:
oof yeah - validates + potentially alters the schedule. i am going to separate it out into `validate` and `correct`

Reply:
899859e should address this. validate now does not modify the schedule. instead we have `correct(mut self) -> Self`, which will modify the schedule.

additionally we now check for durations greater than the maximum block number